File: youtube/lat9-cancer/try2.py
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns 
from statsmodels.stats.outliers_influence import variance_inflation_factor
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.feature_selection import RFE
import statsmodels.api as sm 
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold, cross_val_score
from sklearn.neural_network import MLPClassifier, MLPRegressor


df = pd.read_csv('ML/youtube/lat9-cancer/data.csv')
print(df.head())
print(df.info())
print(df.describe())
print(df.isnull().sum())

df = df.drop('id', axis = 1)
del df['Unnamed: 32']

for col in df:
    print(col)
    print(df[col].nunique())
    print()

print(df['diagnosis'].unique())
df['diagnosis'] = df['diagnosis'].map({'M': 0, 'B': 1})


# fig, ax = plt.subplots(6, 5, figsize=(20, 20))
# # Menggunakan range dari 0 sampai 5 untuk iterasi setiap kolom dalam DataFrame
# sns.scatterplot(x=df.index, y=df['diagnosis'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['radius_mean'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=df.index, y=df['texture_mean'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['perimeter_mean'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=df.index, y=df['area_mean'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=df.index, y=df['smoothness_mean'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=df.index, y=df['compactness_mean'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=df.index, y=df['concavity_mean'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=df.index, y=df['concave points_mean'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=df.index, y=df['symmetry_mean'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=df.index, y=df['fractal_dimension_mean'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=df.index, y=df['radius_se'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=df.index, y=df['texture_se'], ax=ax[2, 2], color='blue')
# sns.scatterplot(x=df.index, y=df['perimeter_se'], ax=ax[2, 3], color='green')
# sns.scatterplot(x=df.index, y=df['area_se'], ax=ax[2, 4], color='orange')
# sns.scatterplot(x=df.index, y=df['smoothness_se'], ax=ax[3, 0], color='red')
# sns.scatterplot(x=df.index, y=df['compactness_se'], ax=ax[3, 1], color='blue')
# sns.scatterplot(x=df.index, y=df['concavity_se'], ax=ax[3, 2], color='green')
# sns.scatterplot(x=df.index, y=df['concave points_se'], ax=ax[3, 3], color='orange')
# sns.scatterplot(x=df.index, y=df['symmetry_se'], ax=ax[3, 4], color='red')
# sns.scatterplot(x=df.index, y=df['fractal_dimension_se'], ax=ax[4, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['radius_worst'], ax=ax[4, 1], color='green')
# sns.scatterplot(x=df.index, y=df['texture_worst'], ax=ax[4, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['perimeter_worst'], ax=ax[4, 3], color='red')
# sns.scatterplot(x=df.index, y=df['area_worst'], ax=ax[4, 4], color='blue')
# sns.scatterplot(x=df.index, y=df['smoothness_worst'], ax=ax[5, 0], color='green')
# sns.scatterplot(x=df.index, y=df['compactness_worst'], ax=ax[5, 1], color='orange')
# sns.scatterplot(x=df.index, y=df['concavity_worst'], ax=ax[5, 2], color='red')
# sns.scatterplot(x=df.index, y=df['concave points_worst'], ax=ax[5, 3], color='blue')
# sns.scatterplot(x=df.index, y=df['symmetry_worst'], ax=ax[5, 4], color='green')
# plt.tight_layout()
# plt.show()

def remove_outlier(df, col):
    # q1 = np.percentile(df[col], 25)
    # q3 = np.percentile(df[col], 75)
    # iqr = q3 - q1
    # df = df[(df[col] >= q1 - 1.5*iqr) & (df[col] <= q3 + 1.5*iqr )]
    
    q1 = df[col].quantile(0.25)
    q3 = df[col].quantile(0.75)
    iqr = q3 - q1
    df = df[(df[col] >= q1 - 1.5*iqr) & (df[col] <= q3 + 1.5*iqr )]
    return df


for col in df:
    df = remove_outlier(df, col)

# fig, ax = plt.subplots(6, 5, figsize=(20, 20))
# # Menggunakan range dari 0 sampai 5 untuk iterasi setiap kolom dalam DataFrame
# sns.scatterplot(x=df.index, y=df['diagnosis'], ax=ax[0, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['radius_mean'], ax=ax[0, 1], color='green')
# sns.scatterplot(x=df.index, y=df['texture_mean'], ax=ax[0, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['perimeter_mean'], ax=ax[0, 3], color='red')
# sns.scatterplot(x=df.index, y=df['area_mean'], ax=ax[0, 4], color='blue')
# sns.scatterplot(x=df.index, y=df['smoothness_mean'], ax=ax[1, 0], color='green')
# sns.scatterplot(x=df.index, y=df['compactness_mean'], ax=ax[1, 1], color='orange')
# sns.scatterplot(x=df.index, y=df['concavity_mean'], ax=ax[1, 2], color='red')
# sns.scatterplot(x=df.index, y=df['concave points_mean'], ax=ax[1, 3], color='blue')
# sns.scatterplot(x=df.index, y=df['symmetry_mean'], ax=ax[1, 4], color='green')
# sns.scatterplot(x=df.index, y=df['fractal_dimension_mean'], ax=ax[2, 0], color='orange')
# sns.scatterplot(x=df.index, y=df['radius_se'], ax=ax[2, 1], color='red')
# sns.scatterplot(x=df.index, y=df['texture_se'], ax=ax[2, 2], color='blue')
# sns.scatterplot(x=df.index, y=df['perimeter_se'], ax=ax[2, 3], color='green')
# sns.scatterplot(x=df.index, y=df['area_se'], ax=ax[2, 4], color='orange')
# sns.scatterplot(x=df.index, y=df['smoothness_se'], ax=ax[3, 0], color='red')
# sns.scatterplot(x=df.index, y=df['compactness_se'], ax=ax[3, 1], color='blue')
# sns.scatterplot(x=df.index, y=df['concavity_se'], ax=ax[3, 2], color='green')
# sns.scatterplot(x=df.index, y=df['concave points_se'], ax=ax[3, 3], color='orange')
# sns.scatterplot(x=df.index, y=df['symmetry_se'], ax=ax[3, 4], color='red')
# sns.scatterplot(x=df.index, y=df['fractal_dimension_se'], ax=ax[4, 0], color='blue')
# sns.scatterplot(x=df.index, y=df['radius_worst'], ax=ax[4, 1], color='green')
# sns.scatterplot(x=df.index, y=df['texture_worst'], ax=ax[4, 2], color='orange')
# sns.scatterplot(x=df.index, y=df['perimeter_worst'], ax=ax[4, 3], color='red')
# sns.scatterplot(x=df.index, y=df['area_worst'], ax=ax[4, 4], color='blue')
# sns.scatterplot(x=df.index, y=df['smoothness_worst'], ax=ax[5, 0], color='green')
# sns.scatterplot(x=df.index, y=df['compactness_worst'], ax=ax[5, 1], color='orange')
# sns.scatterplot(x=df.index, y=df['concavity_worst'], ax=ax[5, 2], color='red')
# sns.scatterplot(x=df.index, y=df['concave points_worst'], ax=ax[5, 3], color='blue')
# sns.scatterplot(x=df.index, y=df['symmetry_worst'], ax=ax[5, 4], color='green')
# plt.tight_layout()
# plt.show()


print(df.head())
numvars = df.drop('diagnosis', axis = 1).columns
scaler = MinMaxScaler()
df[numvars] = scaler.fit_transform(df[numvars])
print(df.head())

df_train, df_test = train_test_split(df, train_size = 0.7, test_size=0.3, random_state=100)
print(df_train.head())
print(df_test.head())

corr = df_train.corr()
print(corr)
# plt.figure(figsize=(16, 10))
# sns.heatmap(corr, annot = True, cmap="YlGnBu")
# plt.show()

x_train = df_train
y_train = df_train.pop('diagnosis')


k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, x_train, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelLR.fit(x_train, y_train)


# sebenarnya tidak dibutuhkan karena jumlah kolomnya sendiri hanya 4
rfe = RFE(estimator=modelLR, n_features_to_select=5)
rfe = rfe.fit(x_train, y_train)
print(rfe.support_)
print(rfe.ranking_)

hasil = pd.DataFrame({
    'features': x_train.columns,
    'support': rfe.support_,
    'rank': rfe.ranking_,
})
print(hasil)

support = x_train.columns[rfe.support_]
not_support = x_train.columns[~rfe.support_]
print(support)
print(not_support)


x_train_rfe = x_train[support]
k_folds = KFold(n_splits=5, shuffle=True, random_state=42)
modelLR  = LogisticRegression()
scoring = 'accuracy'
score = cross_val_score(modelLR, x_train_rfe, y_train, cv = k_folds, n_jobs=1, scoring=scoring)
print(score)
print(round(score.mean(), 2))
modelLR.fit(x_train_rfe, y_train)


vif = pd.DataFrame()
x = x_train_rfe
print(x.shape)

vif['features'] = x.columns
vif['VIF'] = [variance_inflation_factor(x.values, i) for i in range(x.shape[1])]
vif = vif.sort_values('VIF', ascending=False)
print(vif)

vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values('VIF', ascending=False)
print(vif)

y_pred = modelLR.predict(x_train_rfe)

df = pd.DataFrame({
    'y_train': y_train,
    'y_pred': y_pred,
})
print(df.head())
print(df[df['y_train'] != df['y_pred']].head())

akurasi = accuracy_score(y_train, y_pred)
print(akurasi)
cm = confusion_matrix(y_train, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()



x_test = df_test
x_test_rfe = x_test[support]
y_test = df_test.pop('diagnosis')
y_pred = modelLR.predict(x_test_rfe)


df = pd.DataFrame({
    'y_test': y_test,
    'y_pred': y_pred,
})
print(df.head())
print(df[df['y_test'] != df['y_pred']])

akurasi = accuracy_score(y_test, y_pred)
print(akurasi)
cm = confusion_matrix(y_test, y_pred)
print(cm)
sns.heatmap(cm, annot=True,  fmt='.0f', cmap=plt.cm.Blues)
plt.xlabel('y_pred')
plt.ylabel('y_test')
plt.title('confusion matrix')
plt.show()

File: youtube/lat9-cancer/try2.r
library(dplyr)
library(ggplot2)
library(stringr)
library(gridExtra)
library(RMySQL)
library(corrplot)
library(caTools)
library(dplyr)
library(fastDummies)
library(zoo)
library(pROC)
library(ROCR)
library(leaps)
library(car)
library(tidyr)
library(plyr)
library(lubridate)
library(randomForest)
library(caret)
library(patchwork)
library(nnet)
library(klaR) # untuk Naive Bayes
library(rpart) # untuk Decision Trees


df <- read.csv('ML/youtube/lat9-cancer/data.csv')
print(head(df))
print(str(df))
print(summary(df))
print(colSums(is.na(df)))

df <- dplyr::select(df, -id, -X)

for (col in colnames(df)){
    print(col)
    print(head(unique(df[[col]])))
    cat('\n\n')
}



print(sapply(df, function(x) length(unique(na.omit(x)))))
print(sapply(df, function(x) length(unique(x[!is.na(x)]))))

# # TIDAK BISA KARENA ADANYA TABRAKAN DENGAN PACKAGE LAINNTA
# library(dplyr)
# library(tidyverse)
# library(ggplot2)
# hasil <- df %>% summarise(across(everything(), ~n_distinct(na.omit(.))))
# print(hasil)
# hasil <- df %>% summarise(across(everything(), ~n_distinct(.[!is.na(.)])))
# print(hasil)


print(unique(df$diagnosis))

# df$diagnosis <- ifelse(df$diagnosis == 'M', 0, 1)
df$diagnosis <- case_when(
    df$diagnosis == 'M' ~ 0,
    .default = 1
)
print(distinct(df, diagnosis))


# Membuat plot untuk setiap kolom
p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = diagnosis), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p15 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p16 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p17 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p18 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p19 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p20 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p21 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p22 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p23 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p24 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p25 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p26 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p27 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p28 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p29 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p30 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p31 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# Menggabungkan plot menggunakan patchwork
plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + 
        p11 + p12 + p13 + p14 + p15 + p16 + p17 + p18 + p19 + p20 + 
        p21 + p22 + p23 + p24 + p25 + p26 + p27 + p28 + p29 + p30 + p31
print(plot)


remove_outlier <- function(df, col){
    # q1 <- quantile(df[[col]], 0.25)
    # q3 <- quantile(df[[col]], 0.75)
    # iqr <- q3 - q1
    # # Mengaplikasikan aturan IQR untuk menyaring data
    # df <- subset(df, df[[col]] >= q1 - 1.5 * iqr & df[[col]] <= q3 + 1.5 * iqr)


    # q1 <- quantile(df[[col]], 0.25)
    # q3 <- quantile(df[[col]], 0.75)
    # iqr <- q3 - q1
    # # Mengaplikasikan aturan IQR untuk menyaring data
    # df <- filter(df, df[[col]] >= q1 - 1.5 * iqr & df[[col]] <= q3 + 1.5 * iqr)


    q1 <- quantile(df[[col]], 0.25)
    q3 <- quantile(df[[col]], 0.75)
    iqr <- q3 - q1
    # Mengaplikasikan aturan IQR untuk menyaring data
    df <- filter(df, df[[col]] >= q1 - 1.5 * iqr, df[[col]] <= q3 + 1.5 * iqr)
    return(df)
}

for (col in colnames(df)){
    df <- remove_outlier(df, col)
}


# Membuat plot untuk setiap kolom
p1 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = diagnosis), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p2 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p3 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p4 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p5 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p6 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p7 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p8 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_mean), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p9 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_mean), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p10 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_mean), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p11 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_mean), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p12 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p13 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p14 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p15 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p16 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p17 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p18 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_se), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p19 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_se), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p20 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_se), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p21 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_se), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p22 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = radius_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p23 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = texture_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p24 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = perimeter_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p25 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = area_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p26 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = smoothness_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p27 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = compactness_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p28 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concavity_worst), color = "red") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p29 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = concave.points_worst), color = "blue") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p30 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = symmetry_worst), color = "green") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()
p31 <- ggplot(df) +
    geom_point(aes(x = rownames(df), y = fractal_dimension_worst), color = "orange") +
    facet_grid(. ~ ., scales = "free") +
    theme_minimal()

# Menggabungkan plot menggunakan patchwork
plot <- p1 + p2 + p3 + p4 + p5 + p6 + p7 + p8 + p9 + p10 + 
        p11 + p12 + p13 + p14 + p15 + p16 + p17 + p18 + p19 + p20 + 
        p21 + p22 + p23 + p24 + p25 + p26 + p27 + p28 + p29 + p30 + p31
print(plot)

print(head(df))
numvars <- colnames(dplyr::select(df, -diagnosis))
scaler <- preProcess(df[numvars], method = "range")
df[numvars] <- predict(scaler, df[numvars])
print(head(df))


## Set random seed
set.seed(42)
## Create train and test indices
split <- sample.split(df, SplitRatio = 0.7)

## Create train and test dataframes
df_train <- df[split,]
df_test <- df[!split,]
# df_train <- subset(df, split == TRUE)
# df_test <- subset(df, split == FALSE)

## Print first few rows of train and test dataframes
print(head(df_train))
print(head(df_test))
print(dim(df))
print(dim(df_train))
print(dim(df_test))


corr <- cor(df_train)
print(corr)
corplot <- corrplot(corr, method = 'color')
print(corplot)


x_train <- subset(df_train, select = -diagnosis)
y_train <- as.factor(df_train$diagnosis)


ctrl <- trainControl(method="cv", number=5)
modelrf <- train(x_train, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Membuat kontrol RFE
# control <- rfeControl(functions = leapsFuncs, method = "cv", number =10)
control <- rfeControl(functions = rfFuncs, method = "cv", number =10)
# Menjalankan RFE
rfe_model <- rfe(x = x_train, y = y_train, sizes = c(3), rfeControl = control)
# Mendapatkan hasil support dan ranking
support <- predictors(rfe_model)
ranking <- varImp(rfe_model, scale = FALSE)

print(support)
print(ranking)

# Membuat dataframe hasil
hasil <- data.frame(
    Features = rownames(ranking),
    Support = rownames(ranking) %in% support,
    Rank = ranking$Overall
)
print(hasil)


support <- hasil$Features[hasil$Support == TRUE & hasil$Rank > 5]
not_support <- hasil$Features[hasil$Support == FALSE | hasil$Rank < 5]
print(support)
print(not_support)

x_train_rfe <- x_train[support]
print(head(x_train_rfe))

modelrf <- train(x_train_rfe, y_train, method="rf", trControl=ctrl)
print('accuracy mean')
print(modelrf$result)
print(mean(modelrf$result$Accuracy))
cat('\n')


# Menghitung VIF
# Gabungkan x_train_rfe dan y_train menjadi satu data frame
data_train <- cbind(x_train_rfe, response = y_train)
# Membuat formula model linier dengan semua fitur dalam x_train_rfe
formula <- as.formula(paste("response ~", paste(names(x_train_rfe), collapse = " + ")))
# Membuat model linier menggunakan formula tersebut
model <- glm(formula, data = data_train, family = binomial)
# Menghitung VIF
vif_values <- vif(model)
print(vif_values)

# Membuat data frame dengan hasil VIF
vif_df <- data.frame(
    Features = names(vif_values),
    VIF = vif_values
)
vif_df <- vif_df[order(-vif_df$VIF), ]

print(vif_df)

y_pred <- predict(modelrf, x_train_rfe)
df_pred <- data.frame(
    y_train = y_train,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_train != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_train)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_train)
print(cm)



x_test <- dplyr::select(df_test, -diagnosis)
x_test_rfe <- x_test[support]
y_test <- as.factor(df_test$diagnosis)

y_pred <- predict(modelrf, x_test_rfe)
df_pred <- data.frame(
    y_test = y_test,
    y_pred = y_pred
)
print(head(df_pred))

df_pred <- filter(df_pred, y_test != y_pred)
print(head(df_pred))

# Calculate accuracy
accuracy <- confusionMatrix(data = y_pred, reference = y_test)$overall['Accuracy']
print(accuracy)

cat('\n\n\n\n')
# Calculate confusion matrix
print('Confusion Matrix')
cm <- confusionMatrix(data = y_pred, reference = y_test)
print(cm)


